### PR TITLE
5066247: Refine the spec of equals() and hashCode() for j.text.Format classes

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR and [CSR](https://bugs.openjdk.org/browse/JDK-8315720) which refines the spec of `equals()` and `hashCode()` in `java.text.Format` related classes.

The current spec for most of these methods is either  "_Overrides <method_name>_" or are incomplete/wrong (i.e. see `ChoiceFormat`).

This fix adjusts the spec to provide a consistent definition for the overridden methods and specify what is being compared/used to generate a hash code value.

For implementations that use at most a few fields, the values are stated, otherwise a more general term is used as a substitution (i.e. see `DecimalFormat`).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8315720](https://bugs.openjdk.org/browse/JDK-8315720) to be approved

### Issues
 * [JDK-5066247](https://bugs.openjdk.org/browse/JDK-5066247): Refine the spec of equals() and hashCode() for j.text.Format classes (**Bug** - P4)
 * [JDK-8315720](https://bugs.openjdk.org/browse/JDK-8315720): Refine the spec of equals() and hashCode() for j.text.Format classes (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [e322412c](https://git.openjdk.org/jdk/pull/15459/files/e322412cc148551e1ebf163ddb48cf159dfa2e58)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15459/head:pull/15459` \
`$ git checkout pull/15459`

Update a local copy of the PR: \
`$ git checkout pull/15459` \
`$ git pull https://git.openjdk.org/jdk.git pull/15459/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15459`

View PR using the GUI difftool: \
`$ git pr show -t 15459`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15459.diff">https://git.openjdk.org/jdk/pull/15459.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15459#issuecomment-1696443213)

### CODE REVIEW

**Comment by jddarcy** [comment]
This class hierarchy is based around a getClass based notion of class identity rather than an instanceof notion of class identity. The drawbacks of doing this are discussed in "Effective Java, 3rd edition" item 10.

Regardless, getClass vs instanceof cannot be changed now for behavioral compatibility. However, this should be better documented so that subclass outside of the JDK could be written to do the right things in their own equals methods.

File: src/java.base/share/classes/java/text/ChoiceFormat.java [hunk 1]
```diff
@@ -508,7 +513,16 @@ public int hashCode() {
     }
 
     /**
-     * Equality comparison between two
+     * Compares the specified object with this {@code ChoiceFormat} for equality.
+     * Returns true if the object is also a {@code ChoiceFormat} and the
+     * two formats would format any value the same.
+     *
+     * @implNote Implementers should be aware that the default implementation does an
+     * equality check using {@code getClass} rather than {@code instanceOf} when
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/ChoiceFormat.java:521]: Since `instanceof` is a Java keyword, I'd expect it not to be camel-cased.

**Review by naotoj — APPROVED** [review]

**Comment by justin-curtis-lu** [comment]
CSR has been filed -> https://bugs.openjdk.org/browse/JDK-8315720

File: src/java.base/share/classes/java/text/CompactNumberFormat.java [hunk 2]
```diff
@@ -217,6 +217,8 @@ public final class CompactNumberFormat extends NumberFormat {
     @java.io.Serial
     private static final long serialVersionUID = 7128367218649234678L;
 
+    // Non-transient / non-static fields should be added to hashCode impl
+
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/CompactNumberFormat.java:221]: Did you mean `equals()`? `hashCode()` may not have all the non-transient fields.

File: src/java.base/share/classes/java/text/CompactNumberFormat.java [hunk 3]
```diff
@@ -2378,8 +2380,8 @@ public boolean equals(Object obj) {
     /**
      * {@return the hash code for this {@code CompactNumberFormat}}
      *
-     * The hash code value is calculated using any number of fields from the
-     * formatting configuration of this {@code CompactNumberFormat} object.
+     * All fields of this class that are non-transient and non-static are used
+     * to calculate the hash code value.
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/CompactNumberFormat.java:2384]: Is this `All` correct?

**Review by jddarcy — APPROVED** [review]

**Pull request closed without merging**